[Intl] Add `EmojiTransliterator` to translate emoji to many locales

lyrixx opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       |
| License       | MIT
| Doc PR        |

Today, I used the AsciiSlugger and I was really surprised that it replaced
my emoji with nothing (a blank string). I dig a bit, and I think the best way
to solve that is to add a EmojiTransliterator, and then wire it (another PR)

Current API:

```php
$tr = EmojiTransliterator::getInstance('en');
$tr->transliterate('a 😺, 🐈‍⬛, and a 🦁 go to 🏞️... 😍 🎉 💛');
// a grinning cat, black cat, and a lion go to national park️... smiling face with heart-eyes party popper yellow heart
```

---

To do that, I built some transliterator rules, and I committed them to the repository.
The sources come from the official data at https://github.com/unicode-org/cldr

lyrixx committed:
[Intl] Add EmojiTransliterator to translate emoji to many locales

stof commented on src/Symfony/Component/Intl/Resources/bin/transliterator-emoji-rules.php:
```diff
@@ -0,0 +1,78 @@
+    public static function downloadSource(string $sourceFilename): void
+    {
+        if (!file_exists($sourceFilename)) {
+            copy('https://www.unicode.org/Public/emoji/15.0/emoji-test.txt', $sourceFilename);
```
this should download the latest emoji data IMO

lyrixx replied:
How to get it? cf: https://www.unicode.org/Public/emoji/

stloyd replied:
I guess it can be done the same way as an update for ICU:
- https://github.com/unicode-org/icu/blob/main/icu4c/source/data/unidata/emoji-sequences.txt
- https://github.com/symfony/symfony/blob/6.2/src/Symfony/Component/Intl/Resources/bin/update-data.php

stof replied:
you can replace the version number with `latest` to be redirected to the latest version (which is `14.0` currently as Emoji 15.0 is not released yet)

lyrixx replied:
I resolved this conversation, see the other comment by @damienalexandre

stof commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,26 @@
+
+    public static function getInstance(): \Transliterator
+    {
+        if (!isset(static::$instance)) {
```
using `static` to access a private static property is a bug. If you extend that class, it will try to access `ChildClass::$instance`, which is not guaranteed to exists. Use `self::$instance` (and probably make the class final as it is not designed for extensibility)

lyrixx replied:
Thanks, Fixed.

damienalexandre commented on src/Symfony/Component/Intl/Resources/bin/transliterator-emoji-rules.php:
```diff
@@ -0,0 +1,78 @@
+    public static function downloadSource(string $sourceFilename): void
+    {
+        if (!file_exists($sourceFilename)) {
+            copy('https://www.unicode.org/Public/emoji/15.0/emoji-test.txt', $sourceFilename);
```
This file is not supposed to hold the reference "translation" of an emoji. It's a complete list for display testing and keyboard implementation. There is no guaranty about the format being `Format: code points; status # emoji name` forever and it only supports English.

You should use CLDR data, like I do in emoji-search: https://github.com/jolicode/emoji-search/blob/master/tools/build-released.php - it's available as JSON and XML too.

Annotations provide names and keywords for Unicode characters.

What you want here is the "[CLDR Short Name](https://unicode.org/emoji/format.html#col-name)", as shown in bold here: https://unicode-org.github.io/cldr-staging/charts/latest/annotations/romance.html

lyrixx replied:
We talk IRL with @damienalexandre, and we decided: 

* to use https://github.com/unicode-org/cldr/blob/main/common/annotations/fr.xml
* and https://github.com/unicode-org/cldr/blob/main/common/annotationsDerived/fr.xml
* using the `tts` entries (text to speach)
* and to build one file per local

Thanks

lyrixx replied:
Implemented 😍

alamirault commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,37 @@
+    {
+        $rulesFilename = __DIR__."/../Resources/data/transliterator/emoji/$locale.txt";
+        if (!file_exists($rulesFilename)) {
+            throw new \RuntimeException(sprintf('The transliteror rules source does not exist for locale "%s".', $locale));
```
```suggestion
            throw new \RuntimeException(sprintf('The transliterator rules source does not exist for locale "%s".', $locale));
```
transliteror -> transliterator

alamirault commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,37 @@
+            throw new \RuntimeException(sprintf('The transliteror rules source does not exist for locale "%s".', $locale));
+        }
+
+        $transliterator = \Transliterator::createFromRules($a = file_get_contents($rulesFilename));
```
`$a` is not used

alamirault commented on src/Symfony/Component/Intl/Tests/Transliterator/EmojiTransliteratorTest.php:
```diff
@@ -0,0 +1,36 @@
+        $tr = EmojiTransliterator::getInstance('fr');
+
+        $this->assertSame(
+            'a grinningcat, blackcat, and a lion go to nationalpark... smilingfacewithheart-eyes partypopper yellowheart',
```
I'm not sure to understand why we have english transliterations with 'fr' instance :thinking:

nicolas-grekas commented:
Did you want to also provide a map for short names that could work both ways?
Typically mapping :sweat: to `:sweat:`?

lyrixx commented:
@alamirault you review the PR a bit too soon :) It was a WIP (I pushed in order to backup my work).

@nicolas-grekas 
> Typically mapping 😓 to `:sweat:` ?

That's already the case, or I missed something ?

```php
    public function testTransliterate()
    {
        $tr = EmojiTransliterator::getInstance('en');

        $this->assertSame(
            'a grinning cat, black cat, and a lion go to national park️... smiling face with heart-eyes party popper yellow heart',
            $tr->transliterate('a 😺, 🐈‍⬛, and a 🦁 go to 🏞️... 😍 🎉 💛')
        );
    }
```

stof commented on src/Symfony/Component/Intl/Resources/emoji/composer.json:
```diff
@@ -0,0 +1,22 @@
+            "type": "package",
+            "package": {
+                "name": "unicode-org/cldr",
+                "version": "master",
```
invalid version. Should be `dev-master` (and even better as `dev-main` to match the branch name)

lyrixx replied:
Fixed

stof commented on src/Symfony/Component/Intl/Resources/emoji/composer.lock:
```diff
@@ -0,0 +1,325 @@
+        },
+        {
+            "name": "unicode-org/cldr",
+            "version": "master",
```
this committed composer.lock does not help much: it does not lock a specific reference of `unicode-org/cldr` as you don't use a VCS driver finding the reference for a branch. You would need to put the reference in the manually-managed repository metadata instead.

Btw, a local setup won't update the package either when running `composer update` locally, saying it already has that version. Managing manually metadata for a dev version is actually a pain.

lyrixx replied:
Fixed, thanks

stof commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,37 @@
+
+    private static function createTransliterator(string $locale): \Transliterator
+    {
+        $rulesFilename = __DIR__."/../Resources/data/transliterator/emoji/$locale.txt";
```
This should probably validate `$locale` to avoid path traversal issues.

lyrixx replied:
Fixed

stof commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,37 @@
+
+        $transliterator = \Transliterator::createFromRules(file_get_contents($rulesFilename));
+        if (!$transliterator) {
+            throw new \RuntimeException('Unable to create EmojiTransliterator instance.');
```
This should probably use `Transliterator::getErrorMessage()` to improve the error reporting

lyrixx replied:
This is not possible, since the `$translator` is null and the method is not static.

https://www.php.net/manual/en/transliterator.geterrormessage.php

stof replied:
So there is no way to know why an instantiation fails ?

lyrixx replied:
I should have read the PHP source code sooner! It would have help me to build the rules 😅.

I added the error message to the exception!

lyrixx commented:
@stof Thanks for the review, I've addressed your comments. 

I think the PR is now ready.

nicolas-grekas commented on src/Symfony/Component/Intl/CHANGELOG.md:
```diff
@@ -1,6 +1,11 @@
+6.2
+---
+
+* Add `EmojiTransliterator` to translate emoji to many locales.
```
miss space at head of line, no final dot :P

nicolas-grekas commented on src/Symfony/Component/Intl/Resources/emoji/composer.lock:
```diff
@@ -0,0 +1,323 @@
+{
+    "_readme": [
+        "This file locks the dependencies of your project to a known state",
```
what about not committing this file? I don't think it's useful to have it in the repo

nicolas-grekas commented on src/Symfony/Component/Intl/Resources/emoji/composer.json:
```diff
@@ -0,0 +1,22 @@
+                "source": {
+                    "type": "git",
+                    "url": "https://github.com/unicode-org/cldr",
+                    "reference": "production/2022-06-29-1740z"
```
is there a way to reference the latest data to make it even easier to update?

stof replied:
not without having a composer.json in the repository or using a manual management of the install instead of using composer. The `type=package` does not work well with a mutable target

nicolas-grekas replied:
got it, then let's add a note about this in the readme

nicolas-grekas commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        $rulesFilename = __DIR__."/../Resources/data/transliterator/emoji/$locale.txt";
+        if (!file_exists($rulesFilename)) {
```
we prefer using is_file when possible

nicolas-grekas commented on src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php:
```diff
@@ -0,0 +1,41 @@
+        }
+
+        $transliterator = \Transliterator::createFromRules(file_get_contents($rulesFilename));
+        if (!$transliterator) {
```
to be merged with the previous line :)

nicolas-grekas reviewed (approved):
Looks nice :)
I just have minor comments.

In a follow up, it'd be great to add two more locales:
- `en_github` sourced from https://api.github.com/emojis
- `en_slack` sourced from https://github.com/iamcal/emoji-data

lyrixx commented:
@nicolas-grekas Thanks for the review. I addressed your comments.

stof commented on src/Symfony/Component/Intl/Resources/emoji/build.php:
```diff
@@ -0,0 +1,128 @@
+                    $errored[] = [
+                        'locale' => $locale,
+                        'emoji' => $emoji,
+                        'name' => $name,
```
should you include `intl_get_error_message()` here to help debugging ?

lyrixx replied:
Good idea!

lyrixx commented:
I pushed a new version

* cleaned / refactored the Builder
* added more tests
* add a Makefile (will be useful late with github and slack)
* changed the way we build the rule: use `>` instead of `<>` because we cannot reverse the text. The priority is important, and we cannot sort the list for `>` and for `<` at the same time. We can build the other list later if needed)

lyrixx commented:
* Fabbot.io is KO, but the PR is OK (there is french content in this PR...)
* appveyor is KO, but it failing on VarDumper, So it's OK

Except theses 2, everything is green 💚

stof commented on src/Symfony/Component/Intl/Resources/emoji/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Update the rules
+
+To update the rules, you need to update the version of `iamcal/emoji-data` and
```
There is no `iamcal/emoji-data` in it.

lyrixx replied:
Good catch. It was a relicat from a previous version of a next PR (wait, what? 😂 ). Now I download directly what I need!

stof commented on src/Symfony/Component/Intl/Resources/emoji/composer.json:
```diff
@@ -0,0 +1,22 @@
+        "php": ">=7.2",
+        "symfony/filesystem": "^6",
+        "symfony/finder": "^6",
+        "unicode-org/cldr": "2022.06.29"
```
you might use `*` here, to avoid duplicating the version set in the package repository. It simplifies updates (and there is a single version anyway)

lyrixx replied:
Fixed, thanks

lyrixx commented:
ping @symfony/mergers

HeahDude commented on src/Symfony/Component/Intl/Resources/emoji/build.php:
```diff
@@ -0,0 +1,135 @@
+        return $emojisCodePoints;
+    }
+
+    public static function buildRules(array $emojisCodePoints): Generator
```
`: iterable` ?

lyrixx replied:
It's internal tooling. Does it change/improve something?

HeahDude replied:
It hides an implementation detail when callers only need to iterate.
Not sure it improves things though, especially in an internal API.

HeahDude commented on src/Symfony/Component/Intl/Resources/emoji/build.php:
```diff
@@ -0,0 +1,135 @@
+            $xpath = new DOMXPath($document);
+            $results = $xpath->query('.//annotation[@type="tts"]');
+
+            foreach ($results as $result) {
```
```suggestion
            foreach ($xpath->query('.//annotation[@type="tts"]') as $result) {
```
or `$results` has a use I'm missing?

lyrixx replied:
not really except readability :)

HeahDude commented on src/Symfony/Component/Intl/Resources/emoji/build.php:
```diff
@@ -0,0 +1,135 @@
+                self::testEmoji($emoji, $locale);
+                $codePointsCount = mb_strlen($emoji);
+                $mapsByLocale[$locale][$codePointsCount][$emoji] = $name;
+            }
```
Why not yielding after the nested `foreach` with a scoped `$maps` variable instead of keeping everything in memory through `$mapsByLocale`?

lyrixx replied:
Because we need to compute the whole $locale space in one shot.
There are two files to parse before being able to "yield" it :)

```php
            ->in([
                __DIR__.'/vendor/unicode-org/cldr/common/annotationsDerived',
                __DIR__.'/vendor/unicode-org/cldr/common/annotations',
            ])
```

HeahDude replied:
I see, thanks!

HeahDude reviewed (approved):
+441 000 lines 😳 💪 👏

nicolas-grekas committed:
Use PHP for emoji maps

nicolas-grekas commented:
I managed to turn the txt files into php, here is the result: https://github.com/lyrixx/symfony/pull/15
Looks like we should do it :)

nicolas-grekas commented:
BTW, looking at the maps, I'm wondering: shouldn't eg `en` be applied also for `en_ca` (after it likely)? `en_ca` looks very short compared to `en`.

nicolas-grekas reviewed (changes requested):
Locking merge until my last comments are resolved.

lyrixx commented:
> BTW, looking at the maps, I'm wondering: shouldn't eg `en` be applied also for `en_ca` (after it likely)? `en_ca` looks very short compared to `en`.

I missed that point! 

What do you recommend? we can merge en into en_ca?

nicolas-grekas commented:
> What do you recommend? we can merge en into en_ca?

that might increase the size of maps, better run the str_replace twice when needed (with the php-based implem)

nicolas-grekas commented:
> we can merge en into en_ca

I went this way actually, not much to merge. The PR on your fork is up to date.

lyrixx commented:
@nicolas-grekas Feel free to merge this PR, and then you can open a new PR to improve the situation if it needs to

fabpot commented:
@lyrixx Let's finish this PR before merging it. I don't see why we would merge something that we know is not finished yet.

nicolas-grekas commented:
Especially when this means bloating the git history :)

lyrixx commented:
So I think the best way It to merge the PR of @nicolas-grekas (in my PR) (I'll do that in few minutes)

So we can discuss about it. But Unfortunately, I don't understand the new code, and I miss some time to dig into it.

lyrixx commented:
I talked a bit with @nicolas-grekas on slack, and we decided to share our thought on the subject here.

Nikos rewrite all the code to be able to rely on Opcache preloading, in order to increase the performance.

To be honest, I don't like this new implementation (that explains my previous^2 comment). So the rest of this comment might be subjective. But I let you judge :)

The previous code was easier to understand, maintain and work with (see exception message for instance) 😄 :
* old: https://github.com/lyrixx/symfony/blob/intl-trans-emoji-back/src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php
* new: https://github.com/lyrixx/symfony/blob/intl-trans-emoji/src/Symfony/Component/Intl/Transliterator/EmojiTransliterator.php

About the performance, I wrote this Unit Test in both branch

```php
    public function testBench()
    {
        $t = microtime(true);

        for ($i=0; $i < 1000; $i++) {
            $tr = EmojiTransliterator::create('en'); // I also tried with this line outside the loop. Same result ($tr is cached internally) 
            $tr->transliterate('un 😺, 🐈‍⬛, et a 🦁 vont au 🏞️');
        }

        dump((microtime(true) - $t));
    }
```

Here are the results:

Previous code: 0.0145s
New code: 0.088s

It looks like the previous code is 6 times faster than the new code when using only one transliterator instance. It's legit to me, since the new code does not use the translitator anymore but strtr. 


But what about many transliterator? To know that, I used the following command that run for all locales

```
./phpunit src/Symfony/Component/Intl/Tests/Transliterator/EmojiTransliteratorTest.php --filter testAllTransliterator --repeat 100
```

old code 00:01.746
new code: 00:02.592

So here the old code is also faster

---

<details>
<summary>php -v</summary>

```
PHP 8.1.7 (cli) (built: Jun 25 2022 08:13:46) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.7, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.7, Copyright (c), by Zend Technologies
    with blackfire v1.77.0~linux-x64-non_zts81, https://blackfire.io, by Blackfire
```
</details>


<details>
<summary>php -i | grep opca</summary>

```
/etc/php/8.1/cli/conf.d/10-opcache.ini,
opcache.blacklist_filename => no value => no value
opcache.consistency_checks => 0 => 0
opcache.dups_fix => Off => Off
opcache.enable => On => On
opcache.enable_cli => On => On
opcache.enable_file_override => Off => Off
opcache.error_log => no value => no value
opcache.file_cache => no value => no value
opcache.file_cache_consistency_checks => On => On
opcache.file_cache_only => Off => Off
opcache.file_update_protection => 2 => 2
opcache.force_restart_timeout => 180 => 180
opcache.huge_code_pages => Off => Off
opcache.interned_strings_buffer => 8 => 8
opcache.jit => tracing => tracing
opcache.jit_bisect_limit => 0 => 0
opcache.jit_blacklist_root_trace => 16 => 16
opcache.jit_blacklist_side_trace => 8 => 8
opcache.jit_buffer_size => 0 => 0
opcache.jit_debug => 0 => 0
opcache.jit_hot_func => 127 => 127
opcache.jit_hot_loop => 64 => 64
opcache.jit_hot_return => 8 => 8
opcache.jit_hot_side_exit => 8 => 8
opcache.jit_max_exit_counters => 8192 => 8192
opcache.jit_max_loop_unrolls => 8 => 8
opcache.jit_max_polymorphic_calls => 2 => 2
opcache.jit_max_recursive_calls => 2 => 2
opcache.jit_max_recursive_returns => 2 => 2
opcache.jit_max_root_traces => 1024 => 1024
opcache.jit_max_side_traces => 128 => 128
opcache.jit_prof_threshold => 0.005 => 0.005
opcache.lockfile_path => /tmp => /tmp
opcache.log_verbosity_level => 1 => 1
opcache.max_accelerated_files => 15000 => 15000
opcache.max_file_size => 0 => 0
opcache.max_wasted_percentage => 5 => 5
opcache.memory_consumption => 128 => 128
opcache.opt_debug_level => 0 => 0
opcache.optimization_level => 0x7FFEBFFF => 0x7FFEBFFF
opcache.preferred_memory_model => no value => no value
opcache.preload => no value => no value
opcache.preload_user => no value => no value
opcache.protect_memory => Off => Off
opcache.record_warnings => Off => Off
opcache.restrict_api => no value => no value
opcache.revalidate_freq => 2 => 2
opcache.revalidate_path => Off => Off
opcache.save_comments => On => On
opcache.use_cwd => On => On
opcache.validate_permission => Off => Off
opcache.validate_root => Off => Off
opcache.validate_timestamps => On => On
```
</details>

nicolas-grekas commented:
Thanks for challenging my proposal @lyrixx
I figured out that using `strtr()` was way faster than using `str_replace()` for the use case.
I also figured out that we can skip creating a slicing transliterator most of the time.
The PHP version is now faster again :)

nicolas-grekas commented:
My bench: `ab -n 100 http://localhost:8080/` with the script below.

Results with 100 transliterations:
- before: 75 req/s
- after: 390 req/s

Results with 1000 transliterations:
- before: 40 req/s
- after: 70 req/s

```php
<?php
  
use Symfony\Component\Intl\Transliterator\EmojiTransliterator;

require __DIR__.'/vendor/autoload.php';

$tr = EmojiTransliterator::create('en');
//$tr = EmojiTransliterator::getInstance('en');

for ($i=0; $i < 100; $i++) {
    $tr->transliterate('un 😺, 🐈<200d>⬛, et a 🦁 vont au 🏞️ ');
}

echo $tr->transliterate('un 😺, 🐈<200d>⬛, et a 🦁 vont au 🏞️ ');
```

ro0NL commented:
how would we link AsciiSlugger + EmojiTransliterator? is it desired?

nicolas-grekas commented:
> how would we link AsciiSlugger + EmojiTransliterator? is it desired?

In a follow up PR, and yes, I think it's desired.

nicolas-grekas commented:
Thank you @lyrixx.
